// @constatic/base/dist/app.js
// src/app.ts
import { version as djsVersion } from "discord.js";
import { styleText } from "node:util";
import { CommandManager } from "./creators/commands/manager.js";
import { EventManager } from "./creators/events/manager.js";
import { ResponderManager } from "./creators/responders/manager.js";
import { baseErrorHandler } from "./error.js";
import { version } from "./version.js";
var isBun = typeof Bun !== "undefined";

class ConstaticApp {
  commands = new CommandManager(this);
  responders = new ResponderManager(this);
  events = new EventManager(this);
  config;
  static "~instance" = null;
  static getInstance() {
    return this["~instance"] ??= new ConstaticApp;
  }
  constructor() {
    this.config = {
      commands: {},
      responders: {},
      events: {},
      errorHandler: baseErrorHandler
    };
  }
  static destroy() {
    this["~instance"] = null;
  }
  setErrorHandler(handler) {
    this.config.errorHandler = handler;
  }
  intro() {
    console.log();
    console.log("%s %s", styleText("blue", "★ Constatic Base"), styleText("dim", version));
    console.log("%s %s | %s %s", styleText("blueBright", "◌ discord.js"), styleText("dim", djsVersion), isBun ? "◌ Bun" : styleText("green", "⬢ Node.js"), styleText("dim", isBun ? Bun.version : process.versions.node));
    console.log();
  }
  printLogs() {
    console.log([
      ...this.commands.logs,
      ...this.responders.logs,
      ...this.events.logs
    ].join(`
`));
  }
}
export {
  ConstaticApp
};
