// @constatic/base/dist/creators/commands/command.js
// src/creators/commands/command.ts
import { ApplicationCommandOptionType, ApplicationCommandType, InteractionContextType } from "discord.js";

class GroupCommandModule {
  command;
  data;
  constructor(command, data) {
    this.command = command;
    this.data = data;
  }
  subcommand(data) {
    data.group ??= this.data.name;
    this.command.subcommand(data);
    return this;
  }
}

class Command {
  data;
  modules = [];
  constructor(data) {
    this.data = data;
    this.data.type ??= ApplicationCommandType.ChatInput;
    if (this.data.type === ApplicationCommandType.ChatInput) {
      this.data.description ??= this.data.name;
      this.data.name = this.data.name.toLowerCase().replaceAll(" ", "");
    }
    if (this.data.name.length > 32) {
      this.data.name = this.data.name.slice(0, 32);
    }
    if (!this.data.contexts) {
      Object.assign(this.data, {
        contexts: [InteractionContextType.Guild]
      });
    }
  }
  group(data) {
    this.modules.push({
      ...data,
      type: ApplicationCommandOptionType.SubcommandGroup
    });
    return new GroupCommandModule(this, data);
  }
  subcommand(data) {
    this.modules.push({
      ...data,
      type: ApplicationCommandOptionType.Subcommand
    });
    return this;
  }
}
export {
  Command
};
