// @magicyan/discord/dist/functions/message.cjs
'use strict';

const core = require('@magicyan/core');

function findMessage(channel) {
  const messages = channel.messages.cache;
  return {
    async all(limit = Infinity) {
      const messages2 = [];
      let lastMessageId;
      while (true) {
        const fetched = await channel.messages.fetch({ limit: 100, before: lastMessageId });
        messages2.push(...fetched.values());
        lastMessageId = fetched.lastKey();
        if (fetched.size < 100 || messages2.length >= limit)
          break;
      }
      if (limit < messages2.length) {
        const sliced = messages2.slice(0, limit);
        messages2.length = 0;
        messages2.push(...sliced);
      }
      return messages2;
    },
    byId(id) {
      return messages.get(id);
    },
    byContent() {
      return {
        equals(content, ignoreCase = false) {
          return ignoreCase ? messages.find((m) => m.content === content) : messages.find((m) => core.equalsIgnoreCase(m.content, content));
        },
        include(content, ignoreCase = false) {
          return ignoreCase ? messages.find((m) => m.content.includes(content)) : messages.find((m) => core.includesIgnoreCase(m.content, content));
        }
      };
    },
    byFilter(filter) {
      return messages.find(filter);
    }
  };
}
function getMessageURLInfo(url) {
  const match = url.match(/^https:\/\/discord\.com\/channels\/(\d+)\/(\d+)\/(\d+)$/);
  if (!match)
    return {};
  const [_, guildId, channelId, messageId] = match;
  return { guildId, channelId, messageId };
}
async function fetchMessageFromURL(gc, url) {
  const { messageId, channelId } = getMessageURLInfo(url);
  if (!messageId)
    return null;
  const channel = await gc.channels.fetch(channelId).catch(() => null);
  if (!channel || !("messages" in channel))
    return null;
  return channel.messages.fetch(messageId).catch(() => null);
}

exports.fetchMessageFromURL = fetchMessageFromURL;
exports.findMessage = findMessage;
exports.getMessageURLInfo = getMessageURLInfo;
