// @magicyan/core/dist/functions/format.mjs
import { isDefined } from './validation.mjs';

function joinBuilder(separator, ...text) {
  return text.flat(Infinity).filter(isDefined).filter((bol) => typeof bol !== "boolean").join(separator);
}
function notFound(value) {
  return value !== null ? value : void 0;
}
function brBuilder(...text) {
  return joinBuilder("\n", ...text);
}
function spaceBuilder(...text) {
  return joinBuilder(" ", ...text);
}
function replaceText(text, replacements) {
  const keys = Object.keys(replacements);
  if (keys.length === 0)
    return text;
  keys.sort((a, b) => b.length - a.length);
  const pattern = new RegExp(
    keys.map((str) => str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|"),
    "g"
  );
  return text.replace(pattern, (match) => {
    const value = replacements[match];
    return typeof value === "function" ? value(match) : value;
  });
}
function capitalize(word, allWords = false) {
  word = word.trim();
  if (!word)
    return word;
  return allWords ? word.split(" ").map((w) => w[0].toUpperCase() + w.slice(1).toLowerCase()).join(" ") : word[0].toUpperCase() + word.slice(1).toLowerCase();
}
function limitText(text, maxLength, endText = "") {
  return text.length >= maxLength ? text.slice(0, maxLength) + endText : text;
}

export { brBuilder, capitalize, joinBuilder, limitText, notFound, replaceText, spaceBuilder };
