// @magicyan/discord/dist/functions/components/container.mjs
import { isDefined } from '@magicyan/core';
import { ContainerBuilder, resolveColor, ComponentType } from 'discord.js';
import { createComponents } from './components.mjs';
import { isMessage } from '../../guards/message.mjs';
import { isActionRowBuilder } from '../../guards/components/row.mjs';
import { isSectionBuilder } from '../../guards/components/section.mjs';
import { isButtonBuilder } from '../../guards/components/button.mjs';
import { isTextDisplayBuilder } from '../../guards/components/textdisplay.mjs';
import { isMediaGalleryBuilder } from '../../guards/components/gallery.mjs';
import { isFileBuilder } from '../../guards/components/file.mjs';
import { isSeparatorBuilder } from '../../guards/components/separator.mjs';

class ContainerPlusBuilder extends ContainerBuilder {
  constructor(data) {
    const extractData = (message) => message.components.filter((v) => v.type === ComponentType.Container).at(data?.fromIndex ?? 0)?.toJSON() ?? {};
    const constructorData = isDefined(data?.from) ? "message" in data.from ? extractData(data.from.message) : isMessage(data.from) ? extractData(data.from) : data.from.toJSON() : {};
    if (isDefined(data?.accentColor)) {
      constructorData.accent_color = resolveColor(
        data.accentColor
      );
    }
    super(constructorData);
    if (isDefined(data?.components) && data.components.length >= 1) {
      this.spliceComponents(
        0,
        0,
        ...createComponents(data.components)
      );
    }
  }
  /**
   * Sets the accent color of the container.
   *
   * If a color is provided, it resolves and sets the accent color accordingly.
   * If no color or `null` is provided, it clears the accent color.
   *
   * @param color - The color to set as the accent color, or `null` to clear it.
   * @returns The current instance for chaining.
   *
   * @example
   * container.setColor("#ff0000"); // Sets the accent color to red.
   * container.setColor(null);      // Clears the accent color.
   */
  setColor(color) {
    return isDefined(color) ? this.setAccentColor(resolveColor(color)) : this.setAccentColor();
  }
  /**
   * Replaces or removes a component at the specified index in the container.
   *
   * If `data` is provided, it replaces the component at the given index with the new component(s).
   * If `null` is provided, it removes the component at that index.
   *
   * @param index - The index of the component to replace or remove.
   * @param data - The new component data to set, or `null` to remove the component.
   * @returns The current instance for chaining.
   *
   * @example
   * container.setComponent(0, new ButtonBuilder({ label: "Click" }));
   * container.setComponent(1, null); // Removes the component at index 1.
   */
  setComponent(index, data) {
    return this._spliceComponents(index, 1, data);
  }
  insertComponent(argA, argB) {
    if (typeof argA === "number") {
      return this._spliceComponents(argA, 0, argB);
    }
    return this._spliceComponents(this.components.length, 0, argA);
  }
  _spliceComponents(index, deleteCount, data) {
    const args = [index, deleteCount];
    if (isDefined(data))
      args.push(...createComponents(data));
    return this.spliceComponents(...args);
  }
  componentAt(index, type) {
    return isDefined(type) ? this.components.filter((builder) => builder.data.type === type).at(index) : this.components.at(index);
  }
  get buttonComponents() {
    return this.components.filter(
      (comp) => isActionRowBuilder(comp, "buttons") || isSectionBuilder(comp)
    ).flatMap((comp) => {
      if (isSectionBuilder(comp)) {
        if (!comp.accessory || !isButtonBuilder(comp.accessory))
          return [];
        return [comp.accessory];
      }
      return comp.components;
    });
  }
  get sectionComponents() {
    return this.components.filter(isSectionBuilder);
  }
  get selectMenuComponents() {
    return this.components.filter(
      (comp) => isActionRowBuilder(comp, "selects")
    ).flatMap((comp) => comp.components);
  }
  get textDisplayComponents() {
    return this.components.filter(isTextDisplayBuilder);
  }
  get actionRowComponents() {
    return this.components.filter(
      (row) => isActionRowBuilder(row, "buttons") || isActionRowBuilder(row, "selects")
    );
  }
  get mediaGalleryComponents() {
    return this.components.filter(isMediaGalleryBuilder);
  }
  get fileComponents() {
    return this.components.filter(isFileBuilder);
  }
  get separatorComponents() {
    return this.components.filter(isSeparatorBuilder);
  }
}
function createContainer(data, ...items) {
  const isContainerData = (value) => typeof value === "object" && isDefined(value) && !Array.isArray(value);
  if (isContainerData(data)) {
    const { array, from } = data;
    if (array) {
      if (isMessage(from)) {
        return from.components.filter((c) => c.type === ComponentType.Container).map((component) => new ContainerPlusBuilder({ from: component }));
      }
      return [new ContainerPlusBuilder(data)];
    }
    return new ContainerPlusBuilder(data);
  }
  return new ContainerPlusBuilder({
    accentColor: data,
    components: items.flat()
  });
}

export { ContainerPlusBuilder, createContainer };
