// @magicyan/discord/dist/functions/components/container.cjs
'use strict';

const core = require('@magicyan/core');
const discord_js = require('discord.js');
const components = require('./components.cjs');
const message = require('../../guards/message.cjs');
const row = require('../../guards/components/row.cjs');
const section = require('../../guards/components/section.cjs');
const button = require('../../guards/components/button.cjs');
const textdisplay = require('../../guards/components/textdisplay.cjs');
const gallery = require('../../guards/components/gallery.cjs');
const file = require('../../guards/components/file.cjs');
const separator = require('../../guards/components/separator.cjs');

class ContainerPlusBuilder extends discord_js.ContainerBuilder {
  constructor(data) {
    const extractData = (message) => message.components.filter((v) => v.type === discord_js.ComponentType.Container).at(data?.fromIndex ?? 0)?.toJSON() ?? {};
    const constructorData = core.isDefined(data?.from) ? "message" in data.from ? extractData(data.from.message) : message.isMessage(data.from) ? extractData(data.from) : data.from.toJSON() : {};
    if (core.isDefined(data?.accentColor)) {
      constructorData.accent_color = discord_js.resolveColor(
        data.accentColor
      );
    }
    super(constructorData);
    if (core.isDefined(data?.components) && data.components.length >= 1) {
      this.spliceComponents(
        0,
        0,
        ...components.createComponents(data.components)
      );
    }
  }
  /**
   * Sets the accent color of the container.
   *
   * If a color is provided, it resolves and sets the accent color accordingly.
   * If no color or `null` is provided, it clears the accent color.
   *
   * @param color - The color to set as the accent color, or `null` to clear it.
   * @returns The current instance for chaining.
   *
   * @example
   * container.setColor("#ff0000"); // Sets the accent color to red.
   * container.setColor(null);      // Clears the accent color.
   */
  setColor(color) {
    return core.isDefined(color) ? this.setAccentColor(discord_js.resolveColor(color)) : this.setAccentColor();
  }
  /**
   * Replaces or removes a component at the specified index in the container.
   *
   * If `data` is provided, it replaces the component at the given index with the new component(s).
   * If `null` is provided, it removes the component at that index.
   *
   * @param index - The index of the component to replace or remove.
   * @param data - The new component data to set, or `null` to remove the component.
   * @returns The current instance for chaining.
   *
   * @example
   * container.setComponent(0, new ButtonBuilder({ label: "Click" }));
   * container.setComponent(1, null); // Removes the component at index 1.
   */
  setComponent(index, data) {
    return this._spliceComponents(index, 1, data);
  }
  insertComponent(argA, argB) {
    if (typeof argA === "number") {
      return this._spliceComponents(argA, 0, argB);
    }
    return this._spliceComponents(this.components.length, 0, argA);
  }
  _spliceComponents(index, deleteCount, data) {
    const args = [index, deleteCount];
    if (core.isDefined(data))
      args.push(...components.createComponents(data));
    return this.spliceComponents(...args);
  }
  componentAt(index, type) {
    return core.isDefined(type) ? this.components.filter((builder) => builder.data.type === type).at(index) : this.components.at(index);
  }
  get buttonComponents() {
    return this.components.filter(
      (comp) => row.isActionRowBuilder(comp, "buttons") || section.isSectionBuilder(comp)
    ).flatMap((comp) => {
      if (section.isSectionBuilder(comp)) {
        if (!comp.accessory || !button.isButtonBuilder(comp.accessory))
          return [];
        return [comp.accessory];
      }
      return comp.components;
    });
  }
  get sectionComponents() {
    return this.components.filter(section.isSectionBuilder);
  }
  get selectMenuComponents() {
    return this.components.filter(
      (comp) => row.isActionRowBuilder(comp, "selects")
    ).flatMap((comp) => comp.components);
  }
  get textDisplayComponents() {
    return this.components.filter(textdisplay.isTextDisplayBuilder);
  }
  get actionRowComponents() {
    return this.components.filter(
      (row$1) => row.isActionRowBuilder(row$1, "buttons") || row.isActionRowBuilder(row$1, "selects")
    );
  }
  get mediaGalleryComponents() {
    return this.components.filter(gallery.isMediaGalleryBuilder);
  }
  get fileComponents() {
    return this.components.filter(file.isFileBuilder);
  }
  get separatorComponents() {
    return this.components.filter(separator.isSeparatorBuilder);
  }
}
function createContainer(data, ...items) {
  const isContainerData = (value) => typeof value === "object" && core.isDefined(value) && !Array.isArray(value);
  if (isContainerData(data)) {
    const { array, from } = data;
    if (array) {
      if (message.isMessage(from)) {
        return from.components.filter((c) => c.type === discord_js.ComponentType.Container).map((component) => new ContainerPlusBuilder({ from: component }));
      }
      return [new ContainerPlusBuilder(data)];
    }
    return new ContainerPlusBuilder(data);
  }
  return new ContainerPlusBuilder({
    accentColor: data,
    components: items.flat()
  });
}

exports.ContainerPlusBuilder = ContainerPlusBuilder;
exports.createContainer = createContainer;
