// @constatic/base/dist/creators/events/manager.js
// src/creators/events/manager.ts
import { Collection } from "discord.js";
import { BaseManager } from "../manager.js";

class EventManager extends BaseManager {
  get config() {
    return this.app.config.events;
  }
  collection = new Collection;
  getEvents(key) {
    if (!this.collection.has(key)) {
      this.collection.set(key, new Map);
    }
    return this.collection.get(key);
  }
  add(event) {
    const events = this.getEvents(event.data.event);
    events.set(event.data.name, event);
  }
  async onEvent(event, args) {
    const { middleware, onError } = this.config;
    let isBlock = false;
    const block = (...selected) => {
      const tags = event.data.tags ?? [];
      if (selected.length === 0) {
        isBlock = tags.length === 0;
        return;
      }
      isBlock = selected.some((tag) => tags.includes(tag));
    };
    const eventData = { name: event.data.name, args };
    if (middleware)
      await middleware(eventData, block);
    if (isBlock)
      return;
    await event.data.run(...args).catch((err) => {
      if (onError) {
        onError(err, eventData);
        return;
      }
      throw err;
    });
    if (event.data.once) {
      this.getEvents(event.data.event)?.delete(event.data.name);
    }
  }
  register(client) {
    const collection = this.collection.filter((_, key) => key !== "clientReady");
    for (const [key, events] of collection.entries()) {
      client.on(key, (...args) => {
        Promise.all(Array.from(events.values()).map((event) => this.onEvent(event, args)));
      });
    }
  }
  async runReady(client) {
    const collection = this.getEvents("clientReady");
    await Promise.all(Array.from(collection.values()).map((event) => this.onEvent(event, [client])));
  }
}
export {
  EventManager
};
