// @magicyan/core/dist/index.cjs
'use strict';

const check = require('./functions/check.cjs');
const convert = require('./functions/convert.cjs');
const format = require('./functions/format.cjs');
const math = require('./functions/math.cjs');
const ms = require('./functions/ms.cjs');
const timers = require('./functions/timers.cjs');
const validation = require('./functions/validation.cjs');
const _with = require('./functions/with.cjs');
const date = require('./functions/date.cjs');
const sleep = require('./functions/sleep.cjs');



exports.equalsIgnoreCase = check.equalsIgnoreCase;
exports.includesIgnoreCase = check.includesIgnoreCase;
exports.hexToRgb = convert.hexToRgb;
exports.rgbToHex = convert.rgbToHex;
exports.brBuilder = format.brBuilder;
exports.capitalize = format.capitalize;
exports.joinBuilder = format.joinBuilder;
exports.limitText = format.limitText;
exports.notFound = format.notFound;
exports.replaceText = format.replaceText;
exports.spaceBuilder = format.spaceBuilder;
exports.parseFloatOrDefault = math.parseFloatOrDefault;
exports.parseIntOrDefault = math.parseIntOrDefault;
exports.random = math.random;
exports.randomNumber = math.randomNumber;
exports.toMs = ms.toMs;
exports.createInterval = timers.createInterval;
exports.createLoopInterval = timers.createLoopInterval;
exports.createTimeout = timers.createTimeout;
exports.isDefined = validation.isDefined;
exports.isEmail = validation.isEmail;
exports.isNumeric = validation.isNumeric;
exports.isPromise = validation.isPromise;
exports.isUrl = validation.isUrl;
exports.withProperties = _with.withProperties;
exports.withTimeout = _with.withTimeout;
exports.createDate = date.createDate;
exports.sleep = sleep.sleep;
