// @magicyan/discord/dist/guards/components/selectmenu.mjs
import { StringSelectMenuBuilder, UserSelectMenuBuilder, RoleSelectMenuBuilder, ChannelSelectMenuBuilder, MentionableSelectMenuBuilder } from 'discord.js';
import { hasConstructor } from '../utils.mjs';

function isStringSelectMenuBuilder(value) {
  return value instanceof StringSelectMenuBuilder || hasConstructor(value) && value.constructor.name === StringSelectMenuBuilder.name;
}
function isUserSelectMenuBuilder(value) {
  return value instanceof UserSelectMenuBuilder || hasConstructor(value) && value.constructor.name === UserSelectMenuBuilder.name;
}
function isRoleSelectMenuBuilder(value) {
  return value instanceof RoleSelectMenuBuilder || hasConstructor(value) && value.constructor.name === RoleSelectMenuBuilder.name;
}
function isChannelSelectMenuBuilder(value) {
  return value instanceof ChannelSelectMenuBuilder || hasConstructor(value) && value.constructor.name === ChannelSelectMenuBuilder.name;
}
function isMentionableSelectMenuBuilder(value) {
  return value instanceof MentionableSelectMenuBuilder || hasConstructor(value) && value.constructor.name === MentionableSelectMenuBuilder.name;
}
function isAnySelectMenuBuilder(value) {
  return isStringSelectMenuBuilder(value) || isUserSelectMenuBuilder(value) || isRoleSelectMenuBuilder(value) || isChannelSelectMenuBuilder(value) || isMentionableSelectMenuBuilder(value);
}

export { isAnySelectMenuBuilder, isChannelSelectMenuBuilder, isMentionableSelectMenuBuilder, isRoleSelectMenuBuilder, isStringSelectMenuBuilder, isUserSelectMenuBuilder };
