// @magicyan/discord/dist/guards/components/selectmenu.cjs
'use strict';

const discord_js = require('discord.js');
const utils = require('../utils.cjs');

function isStringSelectMenuBuilder(value) {
  return value instanceof discord_js.StringSelectMenuBuilder || utils.hasConstructor(value) && value.constructor.name === discord_js.StringSelectMenuBuilder.name;
}
function isUserSelectMenuBuilder(value) {
  return value instanceof discord_js.UserSelectMenuBuilder || utils.hasConstructor(value) && value.constructor.name === discord_js.UserSelectMenuBuilder.name;
}
function isRoleSelectMenuBuilder(value) {
  return value instanceof discord_js.RoleSelectMenuBuilder || utils.hasConstructor(value) && value.constructor.name === discord_js.RoleSelectMenuBuilder.name;
}
function isChannelSelectMenuBuilder(value) {
  return value instanceof discord_js.ChannelSelectMenuBuilder || utils.hasConstructor(value) && value.constructor.name === discord_js.ChannelSelectMenuBuilder.name;
}
function isMentionableSelectMenuBuilder(value) {
  return value instanceof discord_js.MentionableSelectMenuBuilder || utils.hasConstructor(value) && value.constructor.name === discord_js.MentionableSelectMenuBuilder.name;
}
function isAnySelectMenuBuilder(value) {
  return isStringSelectMenuBuilder(value) || isUserSelectMenuBuilder(value) || isRoleSelectMenuBuilder(value) || isChannelSelectMenuBuilder(value) || isMentionableSelectMenuBuilder(value);
}

exports.isAnySelectMenuBuilder = isAnySelectMenuBuilder;
exports.isChannelSelectMenuBuilder = isChannelSelectMenuBuilder;
exports.isMentionableSelectMenuBuilder = isMentionableSelectMenuBuilder;
exports.isRoleSelectMenuBuilder = isRoleSelectMenuBuilder;
exports.isStringSelectMenuBuilder = isStringSelectMenuBuilder;
exports.isUserSelectMenuBuilder = isUserSelectMenuBuilder;
